/home/adamsorbie/microbial_comp_gen/comparative_genomics/data/trees/input
#!/bin/bash

# this script aligns query sequences (OTU seqs) to the reference alignment generated by ssu-align using papara and then places these seqs in a reference tree using epa-ng

set -e
eval "$(conda shell.bash hook)"
conda activate bioinfo

otu_seqs="~/microbial_comp_gen/comparative_genomics/data/seqs/OTU_seqs/otu_seqs_out/"
ref_alignment_dir="~/microbial_comp_gen/comparative_genomics/seqs/ref_aligmments"
ref_tree_dir="/home/adamsorbie/microbial_comp_gen/comparitive_genomics/data/trees/input"

# write for loop or find here to get ref alignment and trees one by one
# also need to convert fasta to phylip
for i in ${ref_alignment_dir}*.fna;
do
    outname=${i%.fna}
    taxname=$(echo $i | cut -f1 -d"_")
    python fasta2phylip.py -f $i -o ${outname}_phy.phylip
    mkdir ${taxname}_placement
    mv ${outname}_phy.phylip ${taxname_placement}
done

# hard code this for now
# split files by taxonomy



# align OTU seqs with reference
for i in $@;
do
    tree=$(find data/trees/ -depth -name ${i}*.nwk)
    msa=$(find *_placement -depth -name ${i}*.phylip)
    # change output
    papara_static_x86_64 -t $tree-s $msa -q query_seqs.fasta -r -n combined_alignment
done


# place in tree using epa-ng

# extra step to remove ref sequences from query alignment
epa-ng --split ref-align.phy papara_alignment.combined-alignment
epa-ng -s ref_align.fna -t ref_tree.nwk -q output of previous step --model raxmlinfofile


